airbyte-integrations/connectors/source-jdbc/src/main/java/io/airbyte/integrations/source/jdbc/IncrementalUtils.java

```diff
@@ -0,0 +1,94 @@
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package io.airbyte.integrations.source.jdbc;
+
+import com.google.common.base.Charsets;
+import io.airbyte.protocol.models.ConfiguredAirbyteStream;
+import io.airbyte.protocol.models.Field.JsonSchemaPrimitive;
+import java.nio.ByteBuffer;
+
+public class IncrementalUtils {
+
+  public static String getCursorField(ConfiguredAirbyteStream stream) {
+    if (stream.getCursorField().size() == 0) {
+      throw new IllegalStateException("No cursor field specified for stream attempting to do incremental.");
+    } else if (stream.getCursorField().size() > 1) {
+      throw new IllegalStateException("JdbcSource does not support composite cursor fields.");
+    } else {
+      return stream.getCursorField().get(0);
+    }
+  }
+
+  public static JsonSchemaPrimitive getCursorType(ConfiguredAirbyteStream stream, String cursorField) {
+    if (stream.getStream().getJsonSchema().get("properties") == null) {
+      throw new IllegalStateException(String.format("No properties found in stream: %s.", stream.getStream().getName()));
+    }
+
+    if (stream.getStream().getJsonSchema().get("properties").get(cursorField) == null) {
+      throw new IllegalStateException(
+          String.format("Could not find cursor field: %s in schema for stream: %s.", cursorField, stream.getStream().getName()));
+    }
+
+    if (stream.getStream().getJsonSchema().get("properties").get(cursorField).get("type") == null) {
+      throw new IllegalStateException(
+          String.format("Could not find cursor type for field: %s in schema for stream: %s.", cursorField, stream.getStream().getName()));
+    }
+
+    return JsonSchemaPrimitive.valueOf(stream.getStream().getJsonSchema().get("properties").get(cursorField).get("type").asText().toUpperCase());
+  }
+
+  // x < 0 mean replace original
+  // x == 0 means keep original
+  // x > 0 means keep original
+  public static int compareCursors(String original, String candidate, JsonSchemaPrimitive type) {
```

Comment:
can this comparison be done DB side?

Reply:
if we can we should. see: https://github.com/airbytehq/airbyte/pull/1172/files#r536510322. i have some performance concerns around it.

Reply:
😏  `select MAX(col1) from ( values ('a'), ('b'), ('c')) as x (col1);`

Reply:
😏😏😏😏😏🤡 🥨 😏   `ctx.select(DSL.max(DSL.field("col1"))).from(DSL.values(DSL.row("1"), DSL.row("3"), DSL.row("2")).as("t", "col1"))`

Reply:
will decide tomorrow if i'm going to do this in this iteration. but it's nifty.

Reply:
going to hold off on this. i think it is low risk for right now and i don't want to put the work in here if it looks like i'm going to replace it with a jdbcsource anyway.